### ISSUE TITLE
Plan crashes when undetermined resource used in templated JSON

### ISSUE THREAD

**Issue by stevenproctor** [grafana/terraform-provider-grafana#246]
### Terraform Version
* Terraform: 1.0.2
* Terraform Grafana Provider: 1.12.0
* Grafana: v6.6.1 (21bf8b71bc)

### Affected Resource(s)
* grafana_dashboard

### Simple HCL
```hcl
terraform {
  required_version = "~>1.0.0"
  experiments      = [module_variable_optional_attrs]

  required_providers {
    grafana = {
      source  = "grafana/grafana"
      version = "1.12.0"
    }
  }
}

provider "grafana" {
}


resource "grafana_dashboard" "dashboard" {
  config_json = templatefile(
    "${path.module}/dashboard.json.tmpl",
    {
      environment     = "test-env"
      dashboard_title = "proctor-test-dashboard-error"
      uid             = "${random_uuid.dashboard_uid.id}-tf"
    }
  )
}

resource "random_uuid" "dashboard_uid" {
  keepers = {
    environment     = "test-env"
    dashboard_title = "proctor-test-dashboard-error"
  }
}
```

Simple JSON Template:
```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "graphTooltip": 0,
  "links": [],
  "refresh": "30s",
  "schemaVersion": 22,
  "style": "dark",
  "tags": [],
  "templating": {
  },
  "time": {
    "from": "now-30d",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ]
  },
  "timezone": "",
  "title": "${ dashboard_title } (${ environment })",
  "uid": "${ uid }"
}
```

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.
```
╷
│ Error: Plugin did not respond
│ 
│   with module.grafana_dashboard.grafana_dashboard.dashboard,
│   on .terraform/modules/grafana_dashboard/modules/grafana-dashboard/main.tf line 25, in resource "grafana_dashboard" "dashboard":
│   25: resource "grafana_dashboard" "dashboard" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).PlanResourceChange call. The plugin logs may contain
│ more details.
╵

Stack trace from the terraform-provider-grafana_v1.12.0 plugin:

panic: Attempted to unmarshal invalid JSON. This unexpectedly got past schema validation.

goroutine 96 [running]:
github.com/grafana/terraform-provider-grafana/grafana.unmarshalDashboardConfigJSON(0xd7d14a, 0x24, 0xc42fe0)
	github.com/grafana/terraform-provider-grafana/grafana/resource_dashboard.go:268 +0x106
github.com/grafana/terraform-provider-grafana/grafana.normalizeDashboardConfigJSON(0xc42fe0, 0xc000335fa0, 0xb, 0xc42fe0)
	github.com/grafana/terraform-provider-grafana/grafana/resource_dashboard.go:299 +0x145
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diffString(0xc000228a80, 0xd6852e, 0xb, 0xc000301680, 0xc000351508, 0xe70db8, 0xc000340200, 0x203000, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:1375 +0x6e3
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diff(0xc000228a80, 0xd6852e, 0xb, 0xc000301680, 0xc00007ba60, 0xe70db8, 0xc000340200, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:957 +0x4e8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.Diff(0xc000228a80, 0xe6d278, 0xc000322a00, 0xc0004fb7a0, 0xc000329e00, 0x0, 0xc1bb80, 0xc0003fcb30, 0x1323100, 0xc42fe0, ...)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:522 +0x219
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).SimpleDiff(0xc00017ab60, 0xe6d278, 0xc000322a00, 0xc0004fb7a0, 0xc000329e00, 0xc1bb80, 0xc0003fcb30, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:506 +0xa5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).PlanResourceChange(0xc00000d380, 0xe6d278, 0xc000322a00, 0xc00033c190, 0xc000322a00, 0xd28320, 0xc000329300)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:693 +0x7c5
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).PlanResourceChange(0xc0004c8ae0, 0xe6d320, 0xc000322a00, 0xc0004fb6c0, 0xc0004c8ae0, 0xc0003293b0, 0xc0004f0ba0)
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:315 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_PlanResourceChange_Handler(0xd28320, 0xc0004c8ae0, 0xe6d320, 0xc0003293b0, 0xc00010b6e0, 0x0, 0xe6d320, 0xc0003293b0, 0xc00004ae00, 0x37a)
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:362 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00024aa80, 0xe74958, 0xc000602600, 0xc000338240, 0xc0002294a0, 0x12e4d88, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.37.0/server.go:1217 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00024aa80, 0xe74958, 0xc000602600, 0xc000338240, 0x0)
	google.golang.org/grpc@v1.37.0/server.go:1540 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0000a8160, 0xc00024aa80, 0xe74958, 0xc000602600, 0xc000338240)
	google.golang.org/grpc@v1.37.0/server.go:878 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.37.0/server.go:876 +0x1fd

Error: The terraform-provider-grafana_v1.12.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

script returned exit code 1
```

### Expected Behavior

I create resources in AWS using Terraform, and as part of that Terraform, I would expect I can template out Grafana Dashboard JSON for those resources to automatically create dashboards for those resources.

To make Grafana play nicely, I need to have unique IDs on Dashboards and Panels, to do this I use the `random_integer` or `random_uuid` resources from Terraform and use `templatefile` to put those generated ids into my Grafana JSON for their ids.

### Steps to Reproduce
  `terraform plan`
or 
  `terraform apply`

### Important Factoids

Same Terraform works when downgrading to v1.10.0.

It looks like the change with the removal of 
```
	if err != nil {
		// The validate function should've taken care of this.
		return ""
	}
```
in `grafana/resource_dashboard.go` (https://github.com/grafana/terraform-provider-grafana/pull/217/files#diff-d3b5fa10be0054bad74d5b2028ea65350bbdb16aad37b35f78d87625b8d23056L182)

no longer swallows the unmarshalling JSON exception, so it now bombs out on the plan if any of the dynamic data changes that is only available during the "Apply" stage, which would likely make the JSON invalid (as the values in the JSON would not yet be determined).

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-224
- GH-217

**Comment by stevenproctor** [comment]
Added a basic simple template that shows the issue.

In reality we are generating panels from templates for a given set of Lambdas, SQS Queues, and other AWS Resources we are creating in Terraform, and iterating over those to create Panels to add to a Dashboard.

We can't use Grafana templating as we are supporting adding alerts to the various panels, so we have a some JSON templates for different panels for different AWS resources, and use `templatefile` to pass in variables for the templates.

Where this crashes is around things only known at "Apply" time, such as the Panel ID, as that needs to be unique (otherwise Grafana collapses the panels with the same ids in the template), so we generate a `random_integer` for the id of each of the lambda/panel-type combinations and use that in our Panel template.

**Comment by pfouad** [comment]
Thanks for fixing this @trotttrotttrott! When can we expect the new release to be cut?

**Comment by trotttrotttrott** [comment]
@pfouad thanks for the reminder. This was just released in 1.13.4 - https://registry.terraform.io/providers/grafana/grafana/latest.
